File: 49-group-anagrams/group-anagrams.cpp
// class Solution {
// public:
//     vector<vector<string>> groupAnagrams(vector<string>& strs) {
//         vector<vector<string>> result;
//         unordered_map<string, vector<string>>mp;
//         for(int i=0;i<strs.size();i++){
//             string st = strs[i];
//             sort(begin(st) , end(st));
//             mp[st].push_back(strs[i]);
//         }
//         for(auto &it : mp){
//             result.push_back(it.second);
//         }
//         return result;
//     }
// };

// 

class Solution {
public:
    string makesort(string word){
        string ans = "";
        vector<int>freqw(26,0);
        for(char &ch : word){
            freqw[ch -'a']++;
        }
        for(int i=0;i<26;i++){
            if(freqw[i] > 0){
                ans +=string(freqw[i] , i + 'a');
            }
        }
        return ans;
    }
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> result;
        unordered_map<string, vector<string>>mp;
        for(int i=0;i<strs.size();i++){
            string st = strs[i];
            string newword = makesort(st);
            mp[newword].push_back(strs[i]);
        }
        for(auto &it : mp){
            result.push_back(it.second);
        }
        return result;
    }
};

File: 49-group-anagrams/group-anagrams.py
# class Solution:
#     def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
#         res: List[List[str]] = []
#         mp: dict[str, List[str]] = defaultdict(list)

#         for i in range(0,len(strs),1):
#             st = strs[i]
#             key = ''.join(sorted(st))
#             mp[key].append(strs[i])

#         for key in mp:
#             res.append(mp[key])
#         return res


# 

class Solution:
    def sortana(self , word: str) -> str:
        freq : List[int] = [0]*26
        res = ""
        for ch in word:
            freq[ord(ch) - ord('a')] += 1
        for i in range(0,26,1):
            if(freq[i] > 0):
                res += chr(i+ ord('a')) * freq[i]
        return res

    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        res: List[List[str]] = []
        mp: dict[str, List[str]] = defaultdict(list)

        for i in range(0,len(strs),1):
            st = strs[i]
            key = self.sortana(st)
            mp[key].append(strs[i])

        for key in mp:
            res.append(mp[key])
        return res
